### PR TITLE
feat: Add webhdfs support

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR add webdhfs support.

Also, this PR bump opendal to 0.27 to address the duplicated deps `backon`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

- Installation and deployment

### Release note

- Added webhdfs support

</details>

### CODE REVIEW

**Comment by wcy-fdu** [comment]
Thanks for your contribution! 🥰

**Review by wcy-fdu — COMMENTED** [review]
LGTM for your changes! 
I'll run some tests for a while and merge this PR later.

**Comment by Xuanwo** [comment]
> I'll run some tests for a while and merge this PR later.

To run this test on existing hdfs deployment, please make sure `dfs.webhdfs.enabled` in `hdfs-site.xml` has been set to `true`.

**Comment by wcy-fdu** [comment]
I can not start RisingWave via `webhdfs` backend, is there any configuration problem? 
```
Error: while parsing a block mapping, did not find expected key at line 145 column 17
```

**Comment by Xuanwo** [comment]
> I can not start RisingWave via `webhdfs` backend, is there any configuration problem?
> 
> ```
> Error: while parsing a block mapping, did not find expected key at line 145 column 17
> ```

Doesn't seem to be an error from OpenDAL or hdfs.

**Comment by wcy-fdu** [comment]
Can I use `webhdfs` just by adding` dfs.webhdfs.enabled == true `to the original hdfs configuration?

**Comment by wcy-fdu** [comment]
Can I use `webhdfs` just by adding` dfs.webhdfs.enabled == true `to the original hdfs configuration?

**Comment by Xuanwo** [comment]
> Can I use `webhdfs` just by adding`dfs.webhdfs.enabled == true`to the original hdfs configuration?

It should work. OpenDAL test webhdfs in the same way: https://github.com/beyondstorage/setup-hdfs/blob/master/src/setup-hdfs.ts#L31-L48

**Comment by wcy-fdu** [comment]
Good news: successfully start RisingWave via `webhdfs`
Bad news: Got an error in compute node

```
2023-02-17T06:04:57.351493Z  WARN risingwave_storage::hummock::compactor::shared_buffer_compact: Shared Buffer Compaction failed with error: ObjectStore failed with IO error Unexpected (permanent) at write, context: { service: webhdfs, path: hummock_001/2.data } => building request, source: invalid format.
  backtrace of `HummockError`:
   0: <risingwave_storage::hummock::error::HummockError as core::convert::From<risingwave_storage::hummock::error::HummockErrorInner>>::from
             at ./src/storage/src/hummock/error.rs:68:10
   1: <T as core::convert::Into<U>>::into
             at /rustc/3984bc5833db8bfb0acc522c9775383e4171f3de/library/core/src/convert/mod.rs:726:9
   2: risingwave_storage::hummock::error::HummockError::object_io_error
             at ./src/storage/src/hummock/error.rs:78:9
   3: core::ops::function::FnOnce::call_once
             at /rustc/3984bc5833db8bfb0acc522c9775383e4171f3de/library/core/src/ops/function.rs:250:5
   4: core::result::Result<T,E>::map_err
             at /rustc/3984bc5833db8bfb0acc522c9775383e4171f3de/library/core/src/result.rs:860:27
   5: risingwave_storage::hummock::sstable_store::SstableStore::put_sst_data::{{closure}}
             at ./src/storage/src/hummock/sstable_store.rs:205:9
   6: <risingwave_storage::hummock::sstable_store::BatchUploadWriter as risingwave_storage::hummock::sstable::writer::SstableWriter>::finish::{{closure}}::{{closure}}
             at ./src/storage/src/hummock/sstable_store.rs:573:17
```

**Comment by Xuanwo** [comment]
Oh, it's an error happened while buidling http request. What's the config are you using?

File: risedev.yml [hunk 1]
```diff
@@ -128,6 +128,21 @@ profile:
...
+  webhdfs:
+    steps:
+      # - use: etcd
+      - use: meta-node
+      - use: compute-node
+      - use: frontend
+      # If you want to use hdfs as storage backend, configure hdfs namenode and root path:
+      - use: opendal
+        engine: webhdfs
+        namenode: "127.0.0.1:9000"
```
Comment (Xuanwo) [inline risedev.yml:140]: My mistake. webhdfs's config is differnt from hdfs, we should specify the endpoint to `http://127.0.0.1:9000`:

https://opendal.databend.rs/opendal/services/struct.Webhdfs.html

Comment (wcy-fdu) [reply risedev.yml:140]: And we need to config `delegation_token` as well?

Comment (wcy-fdu) [reply risedev.yml:140]: And we need to config `delegation_token` as well?

Comment (Xuanwo) [reply risedev.yml:140]: > And we need to config `delegation_token` as well?

Not need for now. Most hdfs users don't setup this. We plan to support kerberos in the future.

Comment (wcy-fdu) [reply risedev.yml:140]: Seems that `hummock_001/2.data` is identified as  invalid path format🤔

Comment (Xuanwo) [reply risedev.yml:140]: > Seems that `hummock_001/2.data` is identified as invalid path formatthinking

No, the path the valid. But `127.0.0.1:9000` seems not a valid http url.

Comment (wcy-fdu) [reply risedev.yml:140]: I have changed the endpoint to `"http://127.0.0.1:9870"`, but still the same error.

Comment (wcy-fdu) [reply risedev.yml:140]: I have changed the endpoint to `"http://127.0.0.1:9870"`, but still the same error.

Comment (wcy-fdu) [reply risedev.yml:140]: Maybe this is because the hadoop environment is not correct, let me take the time to check it.

Comment (wcy-fdu) [reply risedev.yml:140]: Maybe this is because the hadoop environment is not correct, let me take the time to check it.

Comment (Xuanwo) [reply risedev.yml:140]: Can you output the debugging of the builder via `debug!("{builder:?}")` in `new_webhdfs_engine`? The error is returned by opendal and not related to hdfs itself.

Comment (wcy-fdu) [reply risedev.yml:140]: Seems that already connected to webhdfs.
```
2023-02-17T06:37:28.784007Z DEBUG opendal::services::webhdfs::backend: building backend: Builder { root: Some("rw"), endpoint: Some("127.0.0.1:9870") }    
2023-02-17T06:37:28.784111Z DEBUG opendal::services::webhdfs::backend: backend use root /rw/    
2023-02-17T06:37:28.784131Z DEBUG opendal::services::webhdfs::backend: backend use endpoint 127.0.0.1:9870    
2023-02-17T06:37:28.793976Z DEBUG opendal::services::webhdfs::backend: checking working directory: /rw/   
```

Comment (wcy-fdu) [reply risedev.yml:140]: Seems that already connected to webhdfs.
```
2023-02-17T06:37:28.784007Z DEBUG opendal::services::webhdfs::backend: building backend: Builder { root: Some("rw"), endpoint: Some("127.0.0.1:9870") }    
2023-02-17T06:37:28.784111Z DEBUG opendal::services::webhdfs::backend: backend use root /rw/    
2023-02-17T06:37:28.784131Z DEBUG opendal::services::webhdfs::backend: backend use endpoint 127.0.0.1:9870    
2023-02-17T06:37:28.793976Z DEBUG opendal::services::webhdfs::backend: checking working directory: /rw/   
```

Comment (Xuanwo) [reply risedev.yml:140]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7718ef31167dd2c2a7062be0fb2c03d7

Comment (Xuanwo) [reply risedev.yml:140]: I got it. It's a bug of OpenDAL that didn't prepend `http` for input endpoint. We will fix it in https://github.com/datafuselabs/opendal/issues/1353. You can workaround it via adding `http://` before endpoint.

Comment (wcy-fdu) [reply risedev.yml:140]: That is indeed a problem. But even if I hardcode the endpoint as `http://127.0.0.1:9870 `,  still the same error. 

```
 DEBUG opendal::services::webhdfs::backend: backend use endpoint http://127.0.0.1:9870  
```

```
 WARN risingwave_storage::hummock::compactor::shared_buffer_compact: Shared Buffer Compaction failed with error: ObjectStore failed with IO error Unexpected (permanent) at write, context: { service: webhdfs, service: webhdfs, path: hummock_001/1.data } => redirection fail, source: missing field `Location` at line 1 column 155.
```

Comment (Xuanwo) [reply risedev.yml:140]: Oh, sorry.

webhdfs does not seem to handle rediection correctly, we will fix this in https://github.com/datafuselabs/opendal/issues/1355.

Comment (ClSlaid) [reply risedev.yml:140]: The `namenode` has a port for HDFS and another port for HTTP, did this service use the same port?
And seems like the HTTP response lacks a `Location` field, what's the version of HDFS? For Hadoop3, an HTTP 200 OK response carrying a `Location` JSON should be given.

Comment (wcy-fdu) [reply risedev.yml:140]: hdfs 2.7.3

Comment (ClSlaid) [reply risedev.yml:140]: > hdfs 2.7.3

Just checked the documentation and found out it does not support the `noredirect` parameter the services relying on. I'll fix this ASAP.

**Comment by Xuanwo** [comment]
cc @wcy-fdu, Hi opendal v0.27.2 has been released, please try again~

**Comment by wcy-fdu** [comment]
> cc @wcy-fdu, Hi opendal v0.27.2 has been released, please try again~

Thanks, let me try it.

**Comment by wcy-fdu** [comment]
> cc @wcy-fdu, Hi opendal v0.27.2 has been released, please try again~

Sorry for the wait, just found that the seem error occurs again when after updating to `v0.27.2`, and I guess it's because my env is incorrect. Let me rebuild a machine for verification.

**Comment by wcy-fdu** [comment]
I still can't write successfully with webhdfs, the hdfs-site.xml:
```
<configuration>
    <property>
        <name>dfs.replication</name>
        <value>1</value>
    </property>
    <property>
        <name>dfs.permissions</name>
        <value>false</value>   
    </property>
    <property>
        <name>dfs.namenode.name.dir</name>
        <value>file:/home/ubuntu/hadoop/tmp/dfs/name</value>
    </property>
    <property>
        <name>dfs.datanode.data.dir</name>
        <value>file:/home/ubuntu/hadoop/tmp/dfs/data</value>
    </property>
    <property>
        <name>dfs.webhdfs.enabled</name>
        <value>true</value>
    </property>
    <property>
        <name>dfs.namenode.http-address</name>
        <value>localhost:9870</value>
    </property>
    <property>
        <name>dfs.secondary.http.address</name>
        <value>localhost:9100</value>
    </property>
</configuration>
```

error log:

```
Caused by:
    statement failed: db error: ERROR: QueryError: internal error: Rpc error: gRPC error (Internal error): Storage error: Hummock error: Other error sync task failed for ObjectStore failed with IO error Unexpected (temporary) at write, context: { called: http_util::Client::send_async, service: webhdfs, path: hummock_001/2.data } => send async request, source: error sending request for url (http://127.0.0.1:9870/webhdfs/v1/risingwave/webhdfs/hummock_001/2.data?op=CREATE&overwrite=true): error trying to connect: tcp connect error: Connection refused (os error 111).
      backtrace of `HummockError`:
       0: <risingwave_storage::hummock::error::HummockError as core::convert::From<risingwave_storage::hummock::error::HummockErrorInner>>::from
                 at ./src/storage/src/hummock/error.rs:68:10
       1: <T as core::convert::Into<U>>::into
                 at /rustc/3984bc5833db8bfb0acc522c9775383e4171f3de/library/core/src/convert/mod.rs:726:9
       2: risingwave_storage::hummock::error::HummockError::object_io_error
                 at ./src/storage/src/hummock/error.rs:78:9
```

can you please help me to have a look? cc @Xuanwo

**Comment by ClSlaid** [comment]
Will 'localhost' open to HTTP requests from foreign machines?

**Comment by Xuanwo** [comment]
"Connection refused" indicates that OpenDAL is unable to establish a connection with this endpoint. If you are running HDFS in Docker or Docker Compose, please ensure that the "dfs.namenode.http-address" is set correctly to Docker's network bridge.

**Comment by wcy-fdu** [comment]
> "Connection refused" indicates that OpenDAL is unable to establish a connection with this endpoint. If you are running HDFS in Docker or Docker Compose, please ensure that the "dfs.namenode.http-address" is set correctly to Docker's network bridge.

I'm running webhdfs in aws ec2, so `dfs.namenode.http-address` should be set to the public ip of this machine?

**Comment by Xuanwo** [comment]
> I'm running webhdfs in aws ec2, so `dfs.namenode.http-address` should be set to the public ip of this machine?

The core question is are you running hdfs and risingwave on the same machine? If not, you need to change it to ec2's public ip.

**Comment by wcy-fdu** [comment]
> > I'm running webhdfs in aws ec2, so `dfs.namenode.http-address` should be set to the public ip of this machine?
> 
> The core question is are you running hdfs and risingwave on the same machine? If not, you need to change it to ec2's public ip.

It is the same machine.

**Comment by Xuanwo** [comment]
For current information, I think it's now related to OpenDAL, please make sure `curl http://127.0.0.1:9870` return valid WebHDFS response.

**Comment by Xuanwo** [comment]
Let's have a disscussion at slack for quick resolve: #risingwave-opendal-wg

**Review by wcy-fdu — APPROVED** [review]
Finally I successfully ran RisingWave on `webhdfs`, thanks for your contribution!

File: risedev.yml [hunk 2]
```diff
@@ -128,6 +127,21 @@ profile:
...
```
Comment (wcy-fdu) [inline risedev.yml:139]: Please help to change namenode `127.0.0.1:9870`

File: src/object_store/src/object/opendal_engine/webhdfs.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+// limitations under the License.
+
+use opendal::services::Webhdfs;
+use opendal::Operator;
+
+use super::{EngineType, OpendalObjectStore};
+use crate::object::ObjectResult;
+
+impl OpendalObjectStore {
+    /// create opendal hdfs engine.
```
Comment (wcy-fdu) [inline src/object_store/src/object/opendal_engine/webhdfs.rs:22]: typo

**Pull request merged**